File: IFDS.m
function [Paths, Object, totalLength, foundPath] = IFDS(rho0, sigma0, loc_final, rt, Wp, Paths, Param, L, Object, weatherMat, dwdx, dwdy)

    % Read the parameters
    simMode = Param.simMode;
    scene =  Param.scene;
    sf = Param.sf;
    targetThresh = Param.targetThresh;
    tsim = Param.tsim;
    dt = Param.dt;
    C = Param.C;
    showDisp = Param.showDisp;
    useOptimizer = Param.useOptimizer;
    delta_g = Param.Rg;
    k = Param.k;
    B_U = Param.B_U;
    B_L = Param.B_L;

    % Initialization
    xd = loc_final(1);
    yd = loc_final(2);
    zd = loc_final(3);

    foundPath = 0;

    switch simMode
        case 1 % Simulate by time

            for t = 1:tsim
                Wp(:,t) = real(Wp(:,t));
                xx = Wp(1,t);
                yy = Wp(2,t);
                zz = Wp(3,t);

                if t>1000
                    break
                end
                Object = create_scene(scene, Object, xx, yy, zz, rt);
                if norm([xx yy zz] - [xd yd zd]) < targetThresh
%                     disp('Target destination reached!')
                    Wp = Wp(:,1:t);
                    Paths{L,rt} = Wp;    % Save into cell array
                    foundPath = 1;
                    break
                else
                    % --------------- Weather constraints ------------
                    if k~=0
                        omega = weatherMat(xx+1, yy+101);
                        dwdx_now = dwdx(xx+1, yy+101);
                        dwdy_now = dwdy(xx+1, yy+101);
    
                        for j = 1:Param.numObj
                            Gm = Object(j).Gamma;
                            dGdx = Object(j).n(1);
                            dGdy = Object(j).n(2);
                            dGdz = Object(j).n(3);
                            dGx_p = dGdx + k*exp( (B_L - omega)/(B_L - B_U) * log((Gm-1)/k +1) ) * ...
                                ( log((Gm-1)/k +1)/(B_L-B_U) * dwdx_now - ((B_L-omega)/((Gm-1+k)*(B_L - B_U))) *dGdx );
                            
                            dGy_p = dGdy + k*exp( (B_L - omega)/(B_L - B_U) * log((Gm-1)/k +1) ) * ...
                                ( log((Gm-1)/k+1)/(B_L-B_U) * dwdy_now - ((B_L-omega)/((Gm-1+k)*(B_L - B_U))) *dGdy );
                            
                            dGz_p = dGdz + k*exp( (B_L - omega)/(B_L - B_U) * log((Gm-1)/k +1) ) * ...
                                ( -((B_L-omega)/((Gm-1+k)*(B_L - B_U))) *dGdz );
 

                            Object(j).Gamma = Object(j).Gamma - k* (exp( (B_L - omega)/(B_L - B_U) * log((Gm-1)/k +1) ) -1);
    
                            Object(j).n = [dGx_p; dGy_p; dGz_p];
                            Object(j).t = [dGy_p; -dGx_p; 0];
                        end
                    end
    
                    [UBar, rho0, sigma0, errFlag] = calc_ubar(xx, yy, zz, xd, yd, zd, ...
                        Object, rho0, sigma0, useOptimizer, delta_g, C, sf, t);
                    if errFlag == 1
                        disp("errFlag")
                        break
                    end
                    Wp(:,t+1) = Wp(:,t) + UBar * dt;
                end
%                 t = t+1;
            end
            Wp = Wp(:,1:t);
            Paths{L,rt} = Wp;    % Save into cell array
            if errFlag == 0 %&& all([Object(:).Gamma] > 0.5)
                foundPath = 1;
            else
                foundPath = 0;
            end

        case 2 % simulate by reaching distance
           
            t = 1;
            while true
                xx = Wp(1,t);
                yy = Wp(2,t);
                zz = Wp(3,t);

                if t>1000
                    break
                end
                Object = create_scene(scene, Object, xx, yy, zz, rt);
                if norm([xx yy zz] - [xd yd zd]) < targetThresh
%                     disp('Target destination reached!')
                    Wp = Wp(:,1:t);
                    Paths{L,rt} = Wp;    % Save into cell array
                    foundPath = 1;
                    break
                else
                    % --------------- Weather constraints ------------
                    if k~=0
                        omega = weatherMat(xx+1, yy+101);
                        dwdx_now = dwdx(xx+1, yy+101);
                        dwdy_now = dwdy(xx+1, yy+101);
    
                        for j = 1:Param.numObj
                            Gm = Object(j).Gamma;
                            dGdx = Object(j).n(1);
                            dGdy = Object(j).n(2);
                            dGdz = Object(j).n(3);
                            dGx_p = dGdx + k*exp( (B_L - omega)/(B_L - B_U) * log((Gm-1)/k +1) ) * ...
                                ( log((Gm-1)/k +1)/(B_L-B_U) * dwdx_now - ((B_L-omega)/((Gm-1+k)*(B_L - B_U))) *dGdx );
                            
                            dGy_p = dGdy + k*exp( (B_L - omega)/(B_L - B_U) * log((Gm-1)/k +1) ) * ...
                                ( log((Gm-1)/k+1)/(B_L-B_U) * dwdy_now - ((B_L-omega)/((Gm-1+k)*(B_L - B_U))) *dGdy );
                            
                            dGz_p = dGdz + k*exp( (B_L - omega)/(B_L - B_U) * log((Gm-1)/k +1) ) * ...
                                ( -((B_L-omega)/((Gm-1+k)*(B_L - B_U))) *dGdz );
 

                            Object(j).Gamma = Object(j).Gamma - k* (exp( (B_L - omega)/(B_L - B_U) * log((Gm-1)/k +1) ) -1);
    
                            Object(j).n = [dGx_p; dGy_p; dGz_p];
                            Object(j).t = [dGy_p; -dGx_p; 0];
                        end
                    end
    
                    [UBar, rho0, sigma0, errFlag] = calc_ubar(xx, yy, zz, xd, yd, zd, ...
                        Object, rho0, sigma0, useOptimizer, delta_g, C, sf, t);
                    if errFlag == 1
                        disp("errFlag!")
                        break
                    end
                    Wp(:,t+1) = Wp(:,t) + UBar * dt;
                end
                t = t+1;
            end
            Wp = Wp(:,1:t);
            Paths{L,rt} = Wp;    % Save into cell array
            if errFlag == 0
                foundPath = 1;
            else
                foundPath = 0;
            end  
    end

    %======================= post-Calculation =============================
    if foundPath == 1
        waypoints = Paths{1,1}';         % Calculate pairwise distances between waypoints
        differences = diff(waypoints);   % the differences between consecutive waypoints
        squaredDistances = sum(differences.^2, 2); 
        
        % Calculate the total path length
        totalLength = sum(sqrt(squaredDistances));
        
        % Display the total path length
        if showDisp
            fprintf('Total path length: %.2f m\n', totalLength);
            fprintf('Total flight time: %.2f s\n', totalLength/C);
        end
    else
        totalLength = 0;
    end

end

function [UBar, rho0, sigma0, errFlag]  = calc_ubar(X, Y, Z, xd, yd, zd, Obj, rho0, sigma0, useOptimizer, delta_g, C, sf, time)

    dist = sqrt((X - xd)^2 + (Y - yd)^2 + (Z - zd)^2);

    u = -[C*(X - xd)/dist, C*(Y - yd)/dist, C*(Z - zd)/dist]';
    
    %% Pre-allocation
    numObj = size(Obj,2);
    Mm = zeros(3);
    sum_w = 0;
    errFlag = 0;

    for j = 1:numObj

        % Reading Gamma for each object
        Gamma = Obj(j).Gamma;
        
        % Unit normal vector and Unit tangential vector
        n = Obj(j).n; 
        t = Obj(j).t;
    
        % Object Distance from UAV
        x0 = Obj(j).origin(1);
        y0 = Obj(j).origin(2);
        z0 = Obj(j).origin(3);

        dist_obj = sqrt((X - x0)^2 + (Y - y0)^2 + (Z - z0)^2);

        % Modular Matrix (Perturbation Matrix
        ntu = n' * u;
        if ntu < 0 || sf == 1
            % ---- optimize the rho0, sigma0 for each object
            if useOptimizer == 2
                if mod(time,5)==0 % optimize every 5 waypoints
                    [rho0, sigma0] = path_opt2(Gamma, n, t, u, dist, dist_obj, rho0, sigma0)
                end
            end
            % ---------------------------------------------------
            
%             if useOptimizer == 0
                % Add Gap Constraint
                Rstar = Obj(j).Rstar;
                rho0_star = log(abs(Gamma))/(log(abs(Gamma - ((Rstar + delta_g)/Rstar)^2 + 1))) * rho0;
                rho = rho0_star * exp(1 - 1/(dist_obj * dist));
%             else
%                 % Without SafeGuard
%                 rho = rho0 * exp(1 - 1/(dist_obj * dist));
%             end

            sigma = sigma0 * exp(1 - 1/(dist_obj * dist));

            M = eye(3) - n*n'/(abs(Gamma)^(1/rho)*(n')*n)...
            + t*n'/(abs(Gamma)^(1/sigma)*norm(t)*norm(n));  % tao is removed for now
        elseif ntu >= 0 && sf == 0
            M = eye(3);
        else
            errFlag = 1;
            UBar = u;
            return
            break
        end  

        % Weight
        w = 1;
        for i = 1:numObj
            if i == j
                continue
            else
                w = w * (Obj(i).Gamma - 1)/...
                    ((Obj(j).Gamma - 1) + (Obj(i).Gamma - 1));
            end
        end
        sum_w = sum_w + w;

        % Saving to Field
        Obj(j).n = n;
        Obj(j).t = t;
        Obj(j).dist = dist_obj;
%         Obj(j).rho = rho;
%         Obj(j).sigma = sigma;
        Obj(j).M  = M;
        Obj(j).w = w;
    
    end

    for j = 1:numObj
        Obj(j).w_tilde = Obj(j).w/sum_w;
        Mm = Mm + Obj(j).w_tilde * Obj(j).M;
    end

    UBar = Mm*u;

    function [rho0, sigma0] = path_opt2(Gamma, n, t, u, dist, dist_obj, rho0, sigma0)
    
        xg = [rho0; sigma0];

        lower_bound_rho = 0.05;  % <0.05 issue started to occur
        lower_bound_sigma = 0;
        
        upper_bound_rho = 2;
        upper_bound_sigma = 1;
        
        % Set up the optimization problem
        problem.objective = @(x) norm_ubar(x(1), x(2), Gamma, n, t, u, dist, dist_obj);
        problem.x0 = xg;
        problem.Aineq = [];
        problem.bineq = [];
        problem.Aeq = [];
        problem.beq = [];
        problem.lb = [lower_bound_rho; lower_bound_sigma];
        problem.ub = [upper_bound_rho; upper_bound_sigma];
        problem.nonlcon = [];
        problem.solver = 'fmincon';  % specify the solver
        problem.options = optimoptions('fmincon', ...
            'Algorithm', 'interior-point', ...   % option2: sqp
            'Display', 'off');
        
        % Call fmincon
        [xOpt, fval, exitflag, output] = fmincon(problem);
%         disp(output)
        rho0 = xOpt(1);
        sigma0 = xOpt(2);

    end

end



function willCollideIdx = CheckCollide(x,y,z, Object, sh_old, sv_old, r_old, rt)

    willCollideIdx = zeros(1,size(Object,2));
    for j = 1:size(Object, 2)
        xo = Object(j).origin(rt, 1);
        yo = Object(j).origin(rt, 2);
        zo = Object(j).origin(rt, 3);

        sigma_h = atan2((yo - y)/(xo - x));
        sigma_v = atan2((zo - z) / ( sqrt((xo-x)^2 + (yo-y)^2)  ));

        if (sigma_h == sh_old) && (sigma_v == sv_old) && (sign(r - r_old) == -1) 
            willCollideIdx(j) = 1; % will collide
        end
    end

end

File: ifds/ifds.py
"""IFDS core algorithm.

Ports the main routine and the ``calc_ubar`` nested helper from ``IFDS.m``.

Public API:

- :func:`calc_ubar`  – modulated velocity for a single UAV position.
- :func:`ifds_step`  – single-step update applied in the outer time loop.
- :func:`run_ifds`   – full path-generation loop for one outer ``rt`` iteration,
  reproducing ``IFDS.m``'s ``case 1`` (by time) and ``case 2`` (by distance).
"""

from __future__ import annotations

from typing import TYPE_CHECKING

import numpy as np

from ifds.enums import OptimizerMode, SimMode
from ifds.optimizer import local_opt_rho_sigma
from ifds.params import Param
from ifds.scenes import build_scene

if TYPE_CHECKING:
    from ifds.objects import Object
    from ifds.weather import WeatherField


def _apply_weather_modification(
    objects: list["Object"],
    xx: float,
    yy: float,
    weather: "WeatherField",
    t_idx: int,
    k: float,
    b_l: float,
    b_u: float,
) -> None:
    """Deform each obstacle's Gamma/n/t by the weather field (port of MATLAB block)."""
    omega, dwdx_now, dwdy_now = weather.sample(t_idx, xx, yy)
    for obj in objects:
        gm = obj.gamma
        dgdx, dgdy, dgdz = obj.n
        denom = (b_l - b_u)
        exp_factor = k * np.exp((b_l - omega) / denom * np.log((gm - 1) / k + 1))
        log_term = np.log((gm - 1) / k + 1) / denom
        scale = (b_l - omega) / ((gm - 1 + k) * denom)
        dgx_p = dgdx + exp_factor * (log_term * dwdx_now - scale * dgdx)
        dgy_p = dgdy + exp_factor * (log_term * dwdy_now - scale * dgdy)
        dgz_p = dgdz + exp_factor * (-scale * dgdz)
        obj.gamma = gm - k * (np.exp((b_l - omega) / denom * np.log((gm - 1) / k + 1)) - 1)
        obj.n = np.array([dgx_p, dgy_p, dgz_p])
        obj.t = np.array([dgy_p, -dgx_p, 0.0])


def calc_ubar(
    pos: np.ndarray,
    target: np.ndarray,
    objects: list["Object"],
    rho0: float,
    sigma0: float,
    use_optimizer: OptimizerMode,
    delta_g: float,
    c_speed: float,
    sf: bool,
    time_step: int,
) -> tuple[np.ndarray, float, float, bool]:
    """Compute the modulated velocity ``UBar``.

    Returns ``(ubar, rho0, sigma0, err_flag)``. ``err_flag`` mirrors MATLAB's
    ``errFlag==1`` escape in ``IFDS.m`` (when all obstacles have ``n'u >= 0`` and
    shape-following is disabled, the unmodified goal-seeking ``u`` is returned).
    """
    x, y, z = pos
    xd, yd, zd = target
    dist = float(np.linalg.norm(pos - target))
    u = -c_speed * (pos - target) / dist
    u = u.reshape(3, 1)

    mm = np.zeros((3, 3))
    sum_w = 0.0
    err_flag = False

    for j, obj in enumerate(objects):
        gamma = obj.gamma
        n = obj.n.reshape(3, 1)
        t = obj.t.reshape(3, 1)
        x0, y0, z0 = obj.origin
        dist_obj = float(np.linalg.norm(np.array([x - x0, y - y0, z - z0])))

        ntu = float((n.T @ u).item())
        if ntu < 0 or sf:
            if use_optimizer == OptimizerMode.LOCAL and time_step % 5 == 0:
                rho0, sigma0 = local_opt_rho_sigma(
                    gamma, n.ravel(), t.ravel(), u.ravel(), dist, dist_obj, rho0, sigma0,
                )
            rstar = obj.rstar
            try:
                rho0_star = (
                    np.log(np.abs(gamma))
                    / np.log(np.abs(gamma - ((rstar + delta_g) / rstar) ** 2 + 1))
                    * rho0
                )
            except (ZeroDivisionError, FloatingPointError):
                rho0_star = rho0
            rho = rho0_star * np.exp(1 - 1 / (dist_obj * dist))
            sigma = sigma0 * np.exp(1 - 1 / (dist_obj * dist))
            M = (
                np.eye(3)
                - (n @ n.T) / (np.abs(gamma) ** (1.0 / rho) * (n.T @ n))
                + (t @ n.T) / (np.abs(gamma) ** (1.0 / sigma) * np.linalg.norm(t) * np.linalg.norm(n))
            )
        elif ntu >= 0 and not sf:
            M = np.eye(3)
        else:
            err_flag = True
            return u.ravel(), rho0, sigma0, err_flag

        # Weight (product over other obstacles)
        w = 1.0
        for i, other in enumerate(objects):
            if i == j:
                continue
            w *= (other.gamma - 1) / ((obj.gamma - 1) + (other.gamma - 1))
        sum_w += w

        obj.dist = dist_obj
        obj.M = M
        obj.w = w

    if sum_w == 0:
        return u.ravel(), rho0, sigma0, err_flag

    for obj in objects:
        obj.w_tilde = obj.w / sum_w
        mm = mm + obj.w_tilde * obj.M

    ubar = (mm @ u).ravel()
    return ubar, rho0, sigma0, err_flag


def ifds_step(
    pos: np.ndarray,
    target: np.ndarray,
    objects: list["Object"],
    rho0: float,
    sigma0: float,
    param: Param,
    time_step: int,
) -> tuple[np.ndarray, float, float, bool]:
    """One outer step: returns ``(next_pos, rho0, sigma0, err_flag)``.

    The caller is responsible for scene building and weather modification.
    """
    ubar, rho0, sigma0, err_flag = calc_ubar(
        pos=pos, target=target, objects=objects, rho0=rho0, sigma0=sigma0,
        use_optimizer=param.use_optimizer, delta_g=param.rg,
        c_speed=param.c, sf=param.sf, time_step=time_step,
    )
    next_pos = pos + ubar * param.dt
    return next_pos, rho0, sigma0, err_flag


def run_ifds(
    rho0: float,
    sigma0: float,
    alpha_deg: float,
    loc_final: np.ndarray,
    rt: int,
    wp: np.ndarray,
    paths: list[list[np.ndarray | None]],
    param: Param,
    line_index: int,
    objects: list["Object"],
    weather: "WeatherField | None",
) -> tuple[list[list[np.ndarray | None]], list["Object"], float, bool]:
    """Full IFDS loop for one outer iteration.

    Returns ``(paths, objects, total_length, found_path)``, matching the MATLAB
    ``[Paths, Object, totalLength, foundPath] = IFDS(...)`` signature.
    """
    xd, yd, zd = loc_final
    target = np.array([xd, yd, zd], dtype=float)
    # MATLAB: case 1 (BY_TIME) runs for tsim iterations. Case 2 (BY_DISTANCE)
    # uses `while true` with a break at t>1000. We allocate a buffer big enough
    # for the distance mode and grow `wp` if the caller's buffer is too small.
    max_steps = param.tsim if param.sim_mode == SimMode.BY_TIME else 1001
    if wp.shape[1] < max_steps + 1:
        wp_big = np.zeros((3, max_steps + 1))
        wp_big[:, : wp.shape[1]] = wp
        wp = wp_big
    found_path = False
    err_flag = False
    t = 0  # index of "current" waypoint

    # Use param.static_weather_index under static env (matches main.m)
    weather_t = rt if param.env == "dynamic" else param.static_weather_index

    if param.sim_mode == SimMode.BY_TIME:
        for t in range(max_steps):
            wp[:, t] = np.real(wp[:, t])
            xx, yy, zz = wp[:, t]
            build_scene(param.scene, objects, xx, yy, zz, rt, alpha_deg)

            if np.linalg.norm(np.array([xx, yy, zz]) - target) < param.target_thresh:
                wp_trunc = wp[:, : t + 1]
                paths[line_index][rt] = wp_trunc.copy()
                found_path = True
                break

            if param.k != 0 and weather is not None:
                _apply_weather_modification(objects, xx, yy, weather, weather_t,
                                            param.k, param.b_l, param.b_u)

            ubar, rho0, sigma0, err_flag = calc_ubar(
                pos=wp[:, t], target=target, objects=objects,
                rho0=rho0, sigma0=sigma0, use_optimizer=param.use_optimizer,
                delta_g=param.rg, c_speed=param.c, sf=param.sf, time_step=t + 1,
            )
            if err_flag:
                break
            wp[:, t + 1] = wp[:, t] + ubar * param.dt

        wp_trunc = wp[:, : t + 1]
        paths[line_index][rt] = wp_trunc.copy()
        found_path = not err_flag and found_path

    else:  # BY_DISTANCE
        t = 0
        while True:
            if t >= max_steps - 1:
                break
            wp[:, t] = np.real(wp[:, t])
            xx, yy, zz = wp[:, t]
            build_scene(param.scene, objects, xx, yy, zz, rt, alpha_deg)

            if np.linalg.norm(np.array([xx, yy, zz]) - target) < param.target_thresh:
                wp_trunc = wp[:, : t + 1]
                paths[line_index][rt] = wp_trunc.copy()
                found_path = True
                break

            if param.k != 0 and weather is not None:
                _apply_weather_modification(objects, xx, yy, weather, weather_t,
                                            param.k, param.b_l, param.b_u)

            ubar, rho0, sigma0, err_flag = calc_ubar(
                pos=wp[:, t], target=target, objects=objects,
                rho0=rho0, sigma0=sigma0, use_optimizer=param.use_optimizer,
                delta_g=param.rg, c_speed=param.c, sf=param.sf, time_step=t + 1,
            )
            wp[:, t + 1] = wp[:, t] + ubar * param.dt
            t += 1
        wp_trunc = wp[:, : t + 1]
        paths[line_index][rt] = wp_trunc.copy()

    # Post-calculation: path length
    if found_path and paths[line_index][rt] is not None:
        waypoints = paths[line_index][rt]
        diffs = np.diff(waypoints, axis=1)
        total_length = float(np.sqrt((diffs**2).sum(axis=0)).sum())
        if param.show_disp:
            print(f"Total path length: {total_length:.2f} m")
            print(f"Total flight time: {total_length / param.c:.2f} s")
    else:
        total_length = 0.0

    return paths, objects, total_length, found_path
